src/config/graph.ts

```diff
@@ -0,0 +1,139 @@
+enum VisitedColor {
+  NONE,
+  RED,
+  GREEN,
+  BLACK,
+  BLUE
+}
+
+export interface Node {
+  id: string;
+  color?: VisitedColor;
+  visitCnt?: number;
+}
+
+export interface Edge {
+  source: Node;
+  sink: Node;
+}
+
+export class EdgeClass<N extends Node, E extends Edge> implements Edge {
+  sink: N
+
+  source: N
+
+  constructor(source: N, sink: N) {
+    this.source = source
+    this.sink = sink
+  }
+}
+
+/**
+ * This DAG implementation is tailored for use in config.
+ * It does not represent a general solution for a DAG.
+ */
+export class DirectedAcyclicGraph<N extends Node, E extends EdgeClass<N, E>> {
+  private _nodes = new Map<string, N>();
+
+  private _edges = new Set<EdgeClass<N, E>>();
+
+  setNode(name: string, node: N): void {
+    this._nodes.set(name, node)
+  }
+
+  setEdge(source: N, sink: N): EdgeClass<N, E> {
+    this.setNode(source.id, source)
+    this.setNode(sink.id, sink)
+
+    const edge = new EdgeClass<N, E>(source, sink)
+    if (this._edges.has(edge)) {
```

Comment:
Will this every return true? `edge` is a new edge instance which won't be in the set. I think you want something like `this._edges.find(edge => edge.source.id === source.id && edge.sink.id === sink.id)`?

Reply:
I will check

Reply:
Nope, always creates a new edge. Nice catch.